packages/shared/stub-components.js

```diff
@@ -40,147 +25,72 @@ function getCoreProperties (component: Component): Object {
-      }
-      if (stubs[stub] === true) {
-        components[stub] = createBlankStub({})
-        return
-      }
-
-      if (componentNeedsCompiling(stubs[stub])) {
-        compileTemplate(stubs[stub])
-      }
-
-      if (originalComponents[stub]) {
-        // Remove cached constructor
-        delete originalComponents[stub]._Ctor
-        if (typeof stubs[stub] === 'string') {
-          components[stub] = createStubFromString(stubs[stub], originalComponents[stub])
-        } else {
-          components[stub] = {
-            ...stubs[stub],
-            name: originalComponents[stub].name
-          }
-        }
-      } else {
-        if (typeof stubs[stub] === 'string') {
-          if (!compileToFunctions) {
-            throwError('vueTemplateCompiler is undefined, you must pass components explicitly if vue-template-compiler is undefined')
-          }
-          components[stub] = {
-            ...compileToFunctions(stubs[stub])
-          }
-        } else {
-          components[stub] = {
-            ...stubs[stub]
-          }
-        }
-      }
-      // ignoreElements does not exist in Vue 2.0.x
-      if (Vue.config.ignoredElements) {
-        Vue.config.ignoredElements.push(stub)
-      }
-    })
+    return createStubFromComponent(stubValue, originalComponent && originalComponent.name)
   }
-  return components
 }
 
-function stubComponents (components: Object, stubbedComponents: Object) {
-  Object.keys(components).forEach(component => {
-    // Remove cached constructor
-    delete components[component]._Ctor
-    if (!components[component].name) {
-      components[component].name = component
-    }
-    stubbedComponents[component] = createBlankStub(components[component])
-
-    // ignoreElements does not exist in Vue 2.0.x
-    if (Vue.config.ignoredElements) {
-      Vue.config.ignoredElements.push(component)
-    }
-  })
-}
-
-export function createComponentStubsForAll (component: Component): Object {
-  const stubbedComponents = {}
-
-  if (component.components) {
-    stubComponents(component.components, stubbedComponents)
+function normalizeStubOptions (components: ?Object, stubOptions: ?Object | Array<string>): Object {
+  if (!stubOptions) {
+    stubOptions = Object.keys(components || {})
   }
-
-  let extended = component.extends
-
-  // Loop through extended component chains to stub all child components
-  while (extended) {
-    if (extended.components) {
-      stubComponents(extended.components, stubbedComponents)
-    }
-    extended = extended.extends
+  if (Array.isArray(stubOptions)) {
+    stubOptions = stubOptions.reduce((object, name) => {
+      object[name] = true
+      return object
+    }, {})
   }
+  return stubOptions
+}
 
-  if (component.extendOptions && component.extendOptions.components) {
-    stubComponents(component.extendOptions.components, stubbedComponents)
-  }
+export function createComponentStubs (components: Object = {}, stubOptions: Object): Object {
```

Comment:
Personally I think public methods should be moved to top of file. Didn't do it to satisfy `git blame` as much as possible

Reply:
I don't think we should move the position of functions. Most files in this project have public functions at the bottom of the files

Reply:
Ok, let it stay like that